tla-assignments/src/main/scala/at/forsyte/apalache/tla/assignments/SmtFreeSymbolicTransitionExtractor.scala

```diff
@@ -0,0 +1,174 @@
+package at.forsyte.apalache.tla.assignments
+
+import at.forsyte.apalache.tla.lir.oper._
+import at.forsyte.apalache.tla.lir.storage.{BodyMap, BodyMapFactory, SourceLocator}
+import at.forsyte.apalache.tla.lir.{LetInEx, NameEx, OperEx, TlaEx}
+import at.forsyte.apalache.tla.lir.transformations.TransformationTracker
+
+
+/**
+  * Performs the complete procedure of finding symbolic transitions from the TLA+ implementation.
+  *
+  * Assignment candidates follow the assignment-before-use paradigm.
+  *
+  * @see [[SymbTransGenerator]]
+  *
+  */
+class SmtFreeSymbolicTransitionExtractor(
+                                          tracker: TransformationTracker,
+                                          sourceLoc: SourceLocator
+                                        ) {
+
+  private def getLocString( ex: TlaEx ) = sourceLoc.sourceOf( ex ).getOrElse( "<[UNKNOWN]>" )
+
+  /** Checks whether expressions, which cannot contain assignments, use unassigned variables */
+  private def throwOnUseBeforeAssignment( unassignedVars: Set[String] ) : TlaEx => Unit = {
+    case ex@OperEx( TlaActionOper.prime, NameEx( name ) ) if unassignedVars.contains( name ) =>
```

Comment:
shall we add a case, where an assignment-like expression (or a manual assignment) appears in an expression that is not supposed to have assignments? I understand that line 26 covers that, but it does so in a subtle way. A direct warning to the user would be more helpful.

Reply:
"Assignment-like" is too broad, since it's not unreasonable to use something like `x' = v` in, say, an ITE condition. I could see doing that for manual assignments though.